### ISSUE TITLE
Recent dosemu2 in -t mode returns escape sequence for Alt+F10 (via ConnectBot)

### ISSUE THREAD

**Issue by ecm-pushbx** [dosemu2/dosemu2#1735]
**Describe the bug**

The Insight debugger that I am working with a bit uses Function keys to access many of its features. One such feature is the local menu, which is specific to each window when selected. It is opened using the Alt+F10 key combo. This [is expected to send a 7100h keypress return](https://hg.pushbx.org/ecm/insight/file/79960cec257e/src/insight.inc#l258). dosemu2 used to send that in `dosemu -t` mode. Now however (Build 2.0pre9-dev-20220707-337-g033e26d08) it sends an escape sequence, as I would expect for `-dumb -td -kt` mode. This is not supported by Insight.

**To Reproduce**

1. Start dosemu2 with `-t`
2. Run keybtest.com
3. Press Alt+F10
4. Press Enter to quit the test program

Expected result:

```
7100
```

Actual result:

```
011B
1A5B "["
0332 "2"
0231 "1"
297E "~"
```

**Attach program/game binaries or provide an URL**

https://pushbx.org/ecm/test/20220717/keybtest.com

**Attach the log**

https://pushbx.org/ecm/test/20220717/2130.log

**A regression?**

Works as expected in Build 2.0pre8-20191102-1332-g9cf2b7560

**Additional info**

Running FreeDOS kernel and FreeCOM.

**Comment by andrewbird** [comment]
It seems to work for me. Perhaps it's a function of the terminal or window manager, I'm using lxterminal / xfwm4?
![Screenshot_2022-07-17_21-31-07](https://user-images.githubusercontent.com/220835/179423841-7c987550-05d5-4da8-90a6-d04ffe23a8be.png)

**Comment by ecm-pushbx** [comment]
I'm using ConnectBot to connect over ssh to a Debian server, running screen and bash and dosemu2. As mentioned this works fine with an older dosemu2, which is why I think the error lies here. Other keys (such as just F10, or arrow keys) still work as expected.

**Comment by stsp** [comment]
Even though Alt-f10 is occupied, I
checked Alt-f3 and Alt-f9 and both
were fine in -t mode.
Try to bisect if you are sure its a
regression.

**Comment by ecm-pushbx** [comment]
Got a lot of this error during testing, probably caused by some bad combo of kernel and the config/autoexec files?

```
emufs: redirector enabled
ERROR: INT21: redirecting [home]/proj/keybtest failed (err = 85)
ERROR: -d is not supported with this freedos version
```

Anyway, I was able to copy the test program to `~/.dosemu/drive_c/` and test it this way. Here's the result:

```
c3b5eb52ac9b2b4b8027b84eb4dc204986155bb3 is the first bad commit
commit c3b5eb52ac9b2b4b8027b84eb4dc204986155bb3
Author: Stas Sergeev <stsp@users.sourceforge.net>
Date:   Wed Dec 30 19:42:39 2020 +0300

    term: properly define double-Esc [fixes #1378]

    Not sure why, other keys are defined with "^^" but for
    Esc we need to use "^[".
    It just works that way.

 src/plugin/term/keyb_slang.c | 7 +------
 1 file changed, 1 insertion(+), 6 deletions(-)
```

**Comment by ecm-pushbx** [comment]
Link to that commit: c3b5eb52ac9b2b4b8027b84eb4dc204986155bb3

**Comment by ecm-pushbx** [comment]
Patch that reverts the bad commit, plus a few lines dropped since that are needed to revert it: https://pushbx.org/ecm/test/20220718/20220718.diff (This makes the test work.)

**Comment by stsp** [comment]
No idea what does this do with double-esc.
Please run cat on terminal and show us
what string it prints on alt-f10.
Please find out if the problem is related
to added or removed lines of the patch.
Please find out if it is specific to alt-f10
or can be reproduced with any alt-fN.

**Comment by ecm-pushbx** [comment]
https://pushbx.org/ecm/test/20220719/good1.diff (same diff as in my prior message):

```
diff --git a/src/plugin/term/keyb_slang.c b/src/plugin/term/keyb_slang.c
index abf9b37ce..f70f48615 100644
--- a/src/plugin/term/keyb_slang.c
+++ b/src/plugin/term/keyb_slang.c
@@ -652,6 +652,7 @@ static int init_slang_keymaps(void)
 {
 	SLKeyMap_List_Type *m;
 	char buf[5];
+	unsigned long esc_scan;
 	char * term;
 	char * kf21;
 
@@ -666,6 +667,10 @@ static int init_slang_keymaps(void)
 		config.term_esc_char = 30;
 	}
 
+	/* escape characters are identity mapped in unicode. */
+	esc_scan = config.term_esc_char;
+	esc_scan |= CTRL_MASK;
+
 	keyb_state.Esc_Char = config.term_esc_char + '@';
 
 	if (keyb_state.The_Normal_KeyMap != NULL)
@@ -756,7 +761,12 @@ static int init_slang_keymaps(void)
 	 * Now add one more for the esc character so that sending it twice sends
 	 * it.
 	 */
-	SLkm_define_key("^[^[", (VOID *) DKY_ESC, m);
+	buf[0] = '^';
+	buf[1] = keyb_state.Esc_Char;
+	buf[2] = '^';
+	buf[3] = keyb_state.Esc_Char;
+	buf[4] = 0;
+	SLkm_define_key(buf, (VOID *) esc_scan, m);
 	if (slang_get_error())
 		return -1;
 
```

https://pushbx.org/ecm/test/20220719/good2.diff :

```
diff --git a/src/plugin/term/keyb_slang.c b/src/plugin/term/keyb_slang.c
index abf9b37ce..26c9cb772 100644
--- a/src/plugin/term/keyb_slang.c
+++ b/src/plugin/term/keyb_slang.c
@@ -652,6 +652,7 @@ static int init_slang_keymaps(void)
 {
 	SLKeyMap_List_Type *m;
 	char buf[5];
+	unsigned long esc_scan;
 	char * term;
 	char * kf21;
 
@@ -666,6 +667,10 @@ static int init_slang_keymaps(void)
 		config.term_esc_char = 30;
 	}
 
+	/* escape characters are identity mapped in unicode. */
+	esc_scan = config.term_esc_char;
+	esc_scan |= CTRL_MASK;
+
 	keyb_state.Esc_Char = config.term_esc_char + '@';
 
 	if (keyb_state.The_Normal_KeyMap != NULL)
@@ -756,7 +761,12 @@ static int init_slang_keymaps(void)
 	 * Now add one more for the esc character so that sending it twice sends
 	 * it.
 	 */
-	SLkm_define_key("^[^[", (VOID *) DKY_ESC, m);
+	buf[0] = '^';
+	buf[1] = keyb_state.Esc_Char;
+	buf[2] = '^';
+	buf[3] = keyb_state.Esc_Char;
+	buf[4] = 0;
+	SLkm_define_key(buf, (VOID *) DKY_ESC, m);
 	if (slang_get_error())
 		return -1;
 
```

https://pushbx.org/ecm/test/20220719/goodtest.txt The returned data from keybtest entering Alt+F1 to Alt+F12 using good2.diff:

```
6800
6900
6A00
6B00
6C00
6D00
6E00
6F00
7000
7100
6800
6900
```

https://pushbx.org/ecm/test/20220719/bad.diff This patch exhibits the same problems as occur without any patch. Replacing the `buf` use by the constant `"^[^["` is what is bad, using `esc_scan` or `DKY_ESC` does not seem to matter:

```
diff --git a/src/plugin/term/keyb_slang.c b/src/plugin/term/keyb_slang.c
index abf9b37ce..2c494a16a 100644
--- a/src/plugin/term/keyb_slang.c
+++ b/src/plugin/term/keyb_slang.c
@@ -652,6 +652,7 @@ static int init_slang_keymaps(void)
 {
 	SLKeyMap_List_Type *m;
 	char buf[5];
+	unsigned long esc_scan;
 	char * term;
 	char * kf21;
 
@@ -666,6 +667,10 @@ static int init_slang_keymaps(void)
 		config.term_esc_char = 30;
 	}
 
+	/* escape characters are identity mapped in unicode. */
+	esc_scan = config.term_esc_char;
+	esc_scan |= CTRL_MASK;
+
 	keyb_state.Esc_Char = config.term_esc_char + '@';
 
 	if (keyb_state.The_Normal_KeyMap != NULL)
@@ -756,7 +761,12 @@ static int init_slang_keymaps(void)
 	 * Now add one more for the esc character so that sending it twice sends
 	 * it.
 	 */
-	SLkm_define_key("^[^[", (VOID *) DKY_ESC, m);
+	buf[0] = '^';
+	buf[1] = keyb_state.Esc_Char;
+	buf[2] = '^';
+	buf[3] = keyb_state.Esc_Char;
+	buf[4] = 0;
+	SLkm_define_key("^[^[", (VOID *) esc_scan, m);
 	if (slang_get_error())
 		return -1;
 
```

https://pushbx.org/ecm/test/20220719/badtest.txt This is what the bad.diff patched dosemu outputs for all of Alt+F1 to Alt+F12:

```
071E
184F "O"
1950 "P"
071E
184F "O"
1051 "Q"
071E
184F "O"
1352 "R"
071E
184F "O"
1F53 "S"
071E
1A5B "["
0231 "1"
0635 "5"
297E "~"
071E
1A5B "["
0231 "1"
0837 "7"
297E "~"
071E
1A5B "["
0231 "1"
0938 "8"
297E "~"
071E
1A5B "["
0231 "1"
0A39 "9"
297E "~"
071E
1A5B "["
0332 "2"
0B30 "0"
297E "~"
071E
1A5B "["
0332 "2"
0231 "1"
297E "~"
071E
1A5B "["
0332 "2"
0433 "3"
297E "~"
071E
1A5B "["
0332 "2"
0534 "4"
297E "~"
```

https://pushbx.org/ecm/test/20220719/cat.txt This is what `cat` displays for F1, F2, F3, F4, F9, F10, and each of those with Alt:

```
$ cat
F1 ^[OP F2 ^[OQ F3 ^[OR F4 ^[OS F9 ^[[20~ F10 ^[[21~
F1 P F2 Q F3 R F4 S F9  F10
Alt-F1 ^[^[OP Alt-F2 ^[^[OQ Alt-F3 ^[^[OR Alt-F4 ^[^[OS Alt-F9 ^[^[[20~ Alt-F10 ^[^[[21~
Alt-F1 P Alt-F2 Q Alt-F3 R Alt-F4 S Alt-F9  Alt-F10
```

Using mate-terminal then ssh then screen then bash then `dosemu -t` from the Debian Linux desktop to connect to the server does seem to work with the bad dosemu too, except I had to disable some global MATE shortcuts, and Alt+F11 and Alt+F12 were used by another application and don't work even now that I have disabled that other application's shortcuts.

**Comment by stsp** [comment]
OK thanks for info.
So I couldn't think up anything
better than to use 3 escs for ESC.

**Comment by ecm-pushbx** [comment]
This seems to work, thank you. However, a single Esc keypress (also in `dosemu -t` using keybtest) never seems to arrive. I have to press Esc twice, and that will send two 011Bh keypresses in succession. Some of Insight's windows and local menus use Esc to close, so if you want to close a menu but remain in the window you need to send one Esc keypress, not two. Should I open a new issue for this?

**Comment by stsp** [comment]
Cant reproduce.
When this started to happen?
Just now?

**Comment by ecm-pushbx** [comment]
No, it seems to happen across several revisions. Don't know of any good ones. It does also happen from the mate-terminal running on the desktop, as well as from ConnectBot.

**Comment by ecm-pushbx** [comment]
Worked around this double Esc problem a bit with https://hg.pushbx.org/ecm/insight/rev/8cdae33e89e5 which allows using F10 to close any local menu. (At least, I think it does.)

**Comment by stsp** [comment]
Its strange that double-esc still
works at all as I've just replaced
it with triple...

**Comment by stsp** [comment]
Try Esc in cat, maybe it would
already react only on double-esc?

**Comment by ecm-pushbx** [comment]
Running `cat` either with ConnectBot or mate-terminal behaves in this way: Each Esc keypress generates one `^[` display, each. It appears to have a small delay possibly. Pressing Esc twice will display `^[^[`.

**Comment by stsp** [comment]
Same here, but with dosemu I can't
reproduce your problem.
And can't reproduce double-esc=esc,
because now triple-esc=esc.
There is something wrong somewhere.

**Comment by ecm-pushbx** [comment]
I don't know what happened but I just tried with today's build of dosemu2 on the desktop system, then on the desktop (mate-terminal) connected by ssh to the server, and then in ConnectBot connected to the server. None of them have the problem now. A single Esc generates a single 011Bh code.

I just found out what the common factor is: screen. The terminal multiplexer program called screen. Running screen either locally (desktop system) or over ssh (mate-terminal or ConnectBot), then running `dosemu -t` in that, ends up with the weird Esc behaviour.

**Comment by stsp** [comment]
Please make sure you can run
`cat` under screen and it sees
the single esc.

**Comment by ecm-pushbx** [comment]
It does see the single Esc.

Tried out working with some settings for screen from those:

1. https://superuser.com/questions/1675761/why-does-gnu-screen-cause-me-to-have-to-press-esc-twice-in-vim
2. https://unix.stackexchange.com/questions/8439/gnu-screen-makes-vim-esc-key-slow
3. https://vi.stackexchange.com/questions/11382/delay-between-hitting-esc-in-insert-mode-and-switching-to-normal-mode-within

So far I just managed to get slightly different undesirable behaviour. Whatever I do, screen seems to either pass no Esc or two of them.
